Add muscle tutorial

carme-hp opened the pull request:
<!-- Please submit your Pull Request to the `develop` branch.

It may help to have a look at the file `CONTRIBUTING.md` for a few hints and guidelines. -->

This would be a potential OpenDiHu tutorial. It has one muscle and three tendons (thus multi-coupling). We have the 4 participans running in OpenDiHu. Potentially, the tendon participants could also run in deal.ii. 

Note that it contains no volume coupling.

carme-hp committed:
Create tutorial structure

carme-hp committed:
Add sample structure for an opendihu participant

carme-hp committed:
Move src code to opendihu-solver folder

carme-hp committed:
Complete opendihu participants

carme-hp committed:
Update README

MakisH commented:
My suggestion here regarding the distribution release: Let's aim to merge this before the distribution, as it would help citing and starting a userbase/contributing community for OpenDiHu. We anyway plan to maintain this.

For the code itself, let's not rush it.

Since this is the first tutorial of this kind, I would like to do a sanity check as well.

carme-hp committed:
Get muscle participant to run

carme-hp committed:
Add check for OPENDIHU_HOME

carme-hp committed:
Rename meshName to preciceMeshName

carme-hp committed:
Fix rename

carme-hp committed:
Fix paths, variables and add scripts

carme-hp committed:
Add remove precice-run to clean.sh

carme-hp committed:
Fix bc errors

carme-hp committed:
Setup exchange directory and remove comments

carme-hp committed:
Add information about OpenDiHu

carme-hp committed:
Add .gitignore

MakisH committed:
Merge branch 'develop' into add-muscle-tutorial

MakisH committed:
Fix Markdown linting issues in muscle tutorial README

MakisH committed:
Fix Markdown linting issues in muscle tutorial README

MakisH committed:
Remove unrelated results image - fix check.sh

MakisH committed:
Mark shell scripts as executable

MakisH committed:
Fix shellcheck errors

MakisH committed:
Format with autopep8

autopep8 --in-place --exit-code --aggressive --ignore=E402 --max-line-length=120 file.py

MakisH committed:
Bump GItHub Action autopep8 from v1 to v2

MakisH committed:
Format more files with autopep8

pip3 install --user autopep8
autopep8 --recursive --in-place --aggressive --exit-code --ignore E402 --max-line-length 120 .

MakisH committed:
Double --aggressive autopep8

MakisH committed:
Double --aggressive autopep8

MakisH committed:
Minor fixes in muscle tutorial README

MakisH committed:
Add a clean_opendihu function, adjust cleaning scripts

MakisH committed:
Rename opendihu-solver to solver-opendihu

MakisH committed:
solver-opendihu scripts: set -e -u

- Exit on errors
- Complain for undefined variables

MakisH committed:
Fix clean_opendihu call to clean_precice_logs

MakisH committed:
Further fixes in solver-opendihu/clean.sh

MakisH committed:
Apply set -e -u to more shell scripts

MakisH committed:
Format precice-config.xml

MakisH commented on muscle-tendon-complex/tendon-top-B-opendihu/variables/variables.py:
```diff
@@ -0,0 +1,122 @@
+# time parameters
+# ---------------
+dt_elasticity = 1      # [ms] time step width for elasticity
+end_time = 20000   # [ms] simulation time
```
Strange that the usual convention in preCICE (`precice-config.xml`) is seconds, but here you declare ms, with the same value. It does not matter for preCICE, as long as every participant uses the same assumption, but I would feel more comfortable if we were consistent with the rest of the tutorials here.

Same, of course, for the `dt` etc.

MakisH commented on muscle-tendon-complex/README.md:
```diff
@@ -0,0 +1,123 @@
+
+   - Download input files for OpenDiHu
+
+      OpenDiHu requires input files hosted in [Zenodo](https://zenodo.org/records/4705982) which include CellML files (containing model equations) and mesh files. Downloading these files is necessary to simulate muscles and/or tendons with OpenDiHu. You can [directly download the necessary files](https://zenodo.org/record/4705982/files/input.tgz?download=1). Extract the files and place them in th ...
```
Do they really need to be downloaded into the `opendihu` source directory? What if that is a system directory? What if it is generally a frozen installation?

Can we move these to a local path, or to an `OPENDIHU_EXAMPLES` directory?

carme-hp replied:
I'm not sure I understand your point.
Would `path/to/opendihu/examples/electrophysiology/` be better? 
Or `OPENDIHU_HOME/examples/electrophysiology/`.

MakisH replied:
No, I mean that it is strange to first get OpenDiHu (imagine in the future a `sudo apt install opendihu`) and then copy the examples into the installation directory (similarly to `sudo apt install opendihu-examples`).

What I suggest is to decouple them from the actual installation (for the tutorial) and be able to use them from any directory.

MakisH replied:
I think that option 2 is more intuitive for the user, while option 1 is better for scripting/automation.

Anything is fine, as long as this path if configurable.

carme-hp replied:
okay, I decided for option 2, because I think this is something I could generally do in OpenDiHu.

MakisH commented on muscle-tendon-complex/tendon-top-B-opendihu/variables/variables.py:
Side-effect of having configuration files in Python script format (see https://github.com/opendihu/opendihu/issues/44): How can I know if all variable names exist / are defined / are not misspelled / are needed?

carme-hp replied:
Fair point, at the moment there's no way to check that.

MakisH commented on muscle-tendon-complex/solver-opendihu/src/tendon-solver.cpp:
```diff
@@ -0,0 +1,22 @@
+#include <Python.h>
+#include <cstdlib>
+#include <iostream>
+
+#include <iostream>
+#include "easylogging++.h"
+
+#include "opendihu.h"
```
```suggestion
#include <iostream>
#include <easylogging++.h>
#include <opendihu.h>
```

- avoid duplicate headers
- use `< >` instead of `" "`, as you are not intending to search local directories first (otherwise, these files would have been in the same directory).

MakisH commented on muscle-tendon-complex/solver-opendihu/src/muscle-solver.cpp:
```diff
@@ -0,0 +1,39 @@
+#include <Python.h>
+#include <cstdlib>
+#include <iostream>
+
+#include <iostream>
+#include "easylogging++.h"
+
+#include "opendihu.h"
```
```suggestion
#include <iostream>
#include <easylogging++.h>
#include <opendihu.h>
```

MakisH commented on muscle-tendon-complex/muscle-opendihu/settings-muscle.py:
These settings/configuration files look pretty long to have been manually created. Are they god-given (cite), were they generated using some tool (which? document in the README), or did you create them manually?

carme-hp replied:
i created them manually (they are based on files that BM also created manually afaik).

MakisH reviewed (commented):
I went ahead and fixed most of the linting issues and some other common requirements for tutorials. In terms of general structure, it looks good.

I have not tried to run it. I will once you tell me it is ready.

I have not reviewed the code of the solvers, as I lack the expertise. I trust that they have been validated elsewhere.

See a few more suggestions and questions to continue.

carme-hp commented on muscle-tendon-complex/README.md:
```diff
@@ -0,0 +1,123 @@
+
+## Why multi-coupling?
+
+This is a case with four participants: the muscle and each tendon. In preCICE, there are two options to [couple more than two participants](https://www.precice.org/configuration-coupling-multi.html). The first option is a composition of bi-coupling schemes, in which we must specify the exchange of data in a participant-to-participant manner. However, such a composition is not suited for combining ...
```
I agree:)

carme-hp committed:
Update muscle-tendon-complex/README.md

Co-authored-by: Gerasimos Chourdakis <chourdak@in.tum.de>

carme-hp committed:
Update muscle-tendon-complex/solver-opendihu/src/tendon-solver.cpp

Co-authored-by: Gerasimos Chourdakis <chourdak@in.tum.de>

carme-hp committed:
Update muscle-tendon-complex/solver-opendihu/src/muscle-solver.cpp

Co-authored-by: Gerasimos Chourdakis <chourdak@in.tum.de>

carme-hp commented on muscle-tendon-complex/.gitignore:
Sorry I don't understand. Which default instructions are you referring to?

MakisH replied:
Sorry, I meant "are all these rules to ignore files that will actually be generated when I try to run the tutorial?".

carme-hp replied:
yes! Some when building and some when running

carme-hp committed:
Explain activation and add citation

carme-hp commented on muscle-tendon-complex/README.md:
```diff
@@ -0,0 +1,123 @@
+
+The muscle and tendon meshes are obtained from patient imaging. The interfaces of the tendons and the muscle do not perfectly match, which is a quite common issue due to the limitations of imaging methods and postprocessing tools. Nonetheless, preCICE coupling methods are robust and can handle meshes that do not perfectly match.
+
+TODO: Explain how is the muscle activated!
```
I tried to make it clear in a new [commit](https://github.com/precice/tutorials/pull/471/commits/76b5a0fe6e150ba616e3140ee56c06d510b31c73).

carme-hp committed:
Use pre-commit

carme-hp committed:
Use pre-commit

carme-hp committed:
Merge branch 'add-muscle-tutorial' of github.com:carme-hp/tutorials into add-muscle-tutorial

carme-hp committed:
Add automatically generated folder to gitignore

MakisH commented on muscle-tendon-complex/README.md:
```diff
@@ -62,12 +62,13 @@ The solver solves the so called "monodomain equation" independently for each fib
+      OpenDiHu requires input files hosted in [Zenodo](https://zenodo.org/records/4705982) which include CellML files (containing model equations) and mesh files. Downloading these files is necessary to simulate muscles and/or tendons with OpenDiHu. You can [directly download the necessary files](https://zenodo.org/record/4705982/files/input.tgz?download=1).
 
-   - Setup `$OPENDIHU_HOME` to your `.bashrc` file
+   - Setup `$OPENDIHU_HOME` and `$OPENDIHU_INPUT_DIR` in your `.bashrc` file
```
Just a general terminology thing: When I see "input files", I understand configuration files for a specific simulation. How do these differ that the files included in this repository?

If they are specific to this tutorial, they could also be downloaded directly into this directory (and specified in the `.gitignore`).

carme-hp replied:
What in opendihu we call input file are not  what I understand as configuration files. They are files that specify meshes, i.e., geometry,  and cellml models, i.e., equations. In the configuration files (the ones included in the tutorial, we give the parameters and user settings to use this files.

MakisH commented:
Me and @carme-hp will work together on aligning OpenDiHu with the requirements (e.g., decouple the installation procedure from installing preCICE automatically). Still, this is not a blocking issue.

@carme-hp please check and comment how outdated this.

As discussed in person, tutorials should generally be lightweight cases and @carme-hp wanted to have a look at making the case faster to show some exciting results.
